telemetry/data_feature_usage.go

```diff
@@ -0,0 +1,165 @@
+var TiFlashUsageTelemetry = struct {
+	MinuteWindow []TiFlashUsageItem
+	Lock         sync.RWMutex
+}{Lock: sync.RWMutex{}}
+
+// TiFlashUsageItem is the TiFlash usage telemetry data struct.
+type TiFlashUsageItem struct {
+	TotalNumbers                   atomic.Uint64 `json:"totalNumbers"`
+	TiFlashPushDownNumbers         atomic.Uint64 `json:"tiFlashPushDownNumbers"`
+	TiFlashExchangePushDownNumbers atomic.Uint64 `json:"tiFlashExchangePushDownNumbers"`
+
+	BeginAt *time.Time `json:"beginAt"`
+}
+
+func getTelemetryFeatureUsageInfo(ctx sessionctx.Context) (*featureUsageInfo, error) {
+	// init
+	usageInfo := featureUsageInfo{
+		CoprCacheUsed:    make([]*CoprCacheUsedWindowItem, 0, UpdateInterval/time.Hour),
+		ClusterIndexUsed: make(map[string]bool),
+		TiFlashUsed:      make([]*TiFlashUsageItem, 0, UpdateInterval/time.Hour),
+	}
+
+	// coprocessor cache
+	CoprocessorCacheTelemetry.Lock.Lock()
+	maxLen := 0
+	for _, window := range CoprocessorCacheTelemetry.MinuteWindow {
+		timeSince := time.Since(*window.BeginAt)
+		if timeSince >= UpdateInterval {
+			continue
+		}
+		maxLen = mathutil.Max(maxLen, int(timeSince/time.Hour))
+		i := maxLen - int(timeSince/time.Hour)
+		if len(usageInfo.CoprCacheUsed) <= i {
+			usageInfo.CoprCacheUsed = append(usageInfo.CoprCacheUsed, &CoprCacheUsedWindowItem{})
+		}
+		usageInfo.CoprCacheUsed[i].P0.Add(window.P0.Load())
+		usageInfo.CoprCacheUsed[i].P1.Add(window.P1.Load())
+		usageInfo.CoprCacheUsed[i].P10.Add(window.P10.Load())
+		usageInfo.CoprCacheUsed[i].P20.Add(window.P20.Load())
+		usageInfo.CoprCacheUsed[i].P40.Add(window.P40.Load())
+		usageInfo.CoprCacheUsed[i].P80.Add(window.P80.Load())
+		usageInfo.CoprCacheUsed[i].P100.Add(window.P100.Load())
+		if usageInfo.CoprCacheUsed[i].BeginAt == nil {
+			usageInfo.CoprCacheUsed[i].BeginAt = window.BeginAt
+		}
+	}
+	CoprocessorCacheTelemetry.Lock.Unlock()
+
+	// TiFlash usage
+	TiFlashUsageTelemetry.Lock.Lock()
+	maxLen = 0
+	for _, window := range TiFlashUsageTelemetry.MinuteWindow {
+		timeSince := time.Since(*window.BeginAt)
+		if timeSince >= UpdateInterval {
+			continue
+		}
+		maxLen = mathutil.Max(maxLen, int(timeSince/time.Hour))
+		i := maxLen - int(timeSince/time.Hour)
+		if len(usageInfo.TiFlashUsed) <= i {
+			usageInfo.TiFlashUsed = append(usageInfo.TiFlashUsed, &TiFlashUsageItem{})
+		}
+		usageInfo.TiFlashUsed[i].TotalNumbers.Add(window.TotalNumbers.Load())
+		usageInfo.TiFlashUsed[i].TiFlashPushDownNumbers.Add(window.TiFlashPushDownNumbers.Load())
+		usageInfo.TiFlashUsed[i].TiFlashExchangePushDownNumbers.Add(window.TiFlashExchangePushDownNumbers.Load())
+		if usageInfo.TiFlashUsed[i].BeginAt == nil {
+			usageInfo.TiFlashUsed[i].BeginAt = window.BeginAt
+		}
+	}
+	TiFlashUsageTelemetry.Lock.Unlock()
+
+	// cluster index
+	exec := ctx.(sqlexec.RestrictedSQLExecutor)
+	stmt, err := exec.ParseWithParams(context.TODO(), `
+		SELECT sha2(TABLE_NAME, 256) name, TIDB_PK_TYPE
+		FROM information_schema.tables
```

Comment:
Will this cause too long time, or too much memory, when there are a lot of tables?

Reply:
`Limit` statement can help it.